Mathlib/GroupTheory/SpecificGroups/Coxeter.lean

```diff
@@ -0,0 +1,377 @@
+/-
+Copyright (c) 2023 Newell Jensen. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Newell Jensen
+-/
+import Mathlib.Data.ENat.Basic
+import Mathlib.Data.Matrix.Notation
+import Mathlib.GroupTheory.OrderOfElement
+import Mathlib.GroupTheory.PresentedGroup
+import Mathlib.LinearAlgebra.Matrix.Symmetric
+
+/-!
+# Coxeter Systems
+
+This file defines Coxeter systems.
+
+It introduces a type class `IsCoxeterGroup` for groups that are isomorphic to a group
+presentation corresponding to a Coxeter matrix which is registered in a Coxeter system.
+
+The finite Coxeter groups are classified as the four infinite families:
+
+* Aₙ, Bₙ, Dₙ, I₂ₘ
+
+And the six exceptional systems:
+
+* E₆, E₇, E₈, F₄, H₃, H₄
+
+## Main definitions
+
+* `Matrix.IsCoxeter` : A matrix `IsCoxeter` if it is a symmetric matrix with ones on
+  the diagonal and off-diagonal elements are greater than or equal to two.
+* `Matrix.CoxeterGroup` : The group presentation corresponding to a Coxeter matrix.
+* `CoxeterSystem` : A structure recording the isomorphism between a group `W` and the group
+  presentation corresponding to a Coxeter matrix, i.e. `Matrix.CoxeterGroup B M`.
+* `IsCoxeterGroup` : A group is a Coxeter group if it is registered in a Coxeter System.
+
+## References
+
+* https://en.wikipedia.org/wiki/Coxeter_group
+
+* [N. Bourbaki, *Lie Groups and Lie Algebras, Chapters 4--6*](bourbaki1968) chapter IV
+  pages 13--15
+
+## Tags
+
+coxeter system, coxeter group
+-/
+
+noncomputable section
+
+variable {B : Type*} [DecidableEq B]
+
+variable (M : Matrix B B ℕ∞)
+
+variable (B)
+
+/-- A matrix `IsCoxeter` if it is a symmetric matrix with ones on the diagonal
+and off-diagonal elements are greater than or equal to two. -/
+class Matrix.IsCoxeter (M : Matrix B B ℕ∞) : Prop where
+  symmetric : M.IsSymm
+  diagonal : ∀ i j : B, i = j → M i j = 1
+  off_diagonal : ∀ i j : B, i ≠ j → 2 ≤ M i j
```

Comment:
```suggestion
  symmetric : M.IsSymm := by aesop
  diagonal : ∀ i j : B, i = j → M i j = 1 := by aesop
  off_diagonal : ∀ i j : B, i ≠ j → 2 ≤ M i j := by aesop
```

Reply:
Actually, `diagonal` should be `∀ i : B, M i i = 1`

Reply:
> Doing so, if the definitions are set up correctly, we do not need to provide any proof as lean will automatically try `aesop`.

I am not sure what you meant by 'not providing any proof' but I assume you mean we wouldn't need to provide proofs for the `instance`s?

I wasn't able to prove `IsCoxeter` by using for `aesop` for the all the matrices I have.  I was able to find proofs, but not using just `aesop`.

Reply:
I pushed what I currently have so at least you can see most of the changes you were mentioning above (still not sure on the `instance`s).

Reply:
If we use  `ℕ` instead of `ℕ+`, for `A_n`, we only need the following code:
```lean
instance AₙIsCoxeter : IsCoxeter (Aₙ n) where
```